### PR TITLE
Add nopush TCP socket option

### PR DESCRIPTION
**Pull request by igors**
Add nopush TCP socket option

This translates to TCP_CORK on Linux and TCP_NOPUSH on
BSD.

In effect, this acts as super-Nagle: no partial TCP segments
are sent out until this option is turned off. Once turned off,
all accumulated unsent data is sent out immediately. The latter
is *not* the case on OSX, hence the implementation ignores
"nopush" on OSX to reduce confusion.

The change is motivated by [ERL-698](https://bugs.erlang.org/browse/ERL-698)

### CODE REVIEW

**Comment by jhogberg** [comment]
Thanks for the PR!

It looks great aside from the lack of a test case which should be fairly simple to add to `inet_sockopt_SUITE`. I don't think we need much more than simple API tests since the option will be exercised as part of the sendfile suite later on.

Regarding OS X, I think the best way forward is to treat the option like it doesn't exist, and include your idea to [wrap sendfile with cork/uncork](https://github.com/igors/otp/commit/8e45a074d7756f226851c26a638d925f752447a3) in this PR.

File: erts/preloaded/src/prim_inet.erl [hunk 1]
```diff
@@ -520,13 +520,34 @@ sendfile(S, FileHandle, Offset, Length)
 sendfile(S, FileHandle, Offset, Length) ->
     case erlang:port_info(S, connected) of
         {connected, Pid} when Pid =:= self() ->
-            sendfile_1(S, FileHandle, Offset, Length);
+            Uncork = sendfile_maybe_cork(S),
+            Result = sendfile_1(S, FileHandle, Offset, Length),
+            sendfile_maybe_uncork(S, Uncork),
```
Comment (jhogberg) [inline erts/preloaded/src/prim_inet.erl]: Dialyzer complains about unmatched return values here and on line 538. I believe it should be safe to suppress these warnings (`_ = ...`) and have done so for the nightly builds.

You can check this yourself with `scripts/run-dialyzer` after rebuilding preloaded modules.

Comment (igors) [reply erts/preloaded/src/prim_inet.erl]: Updated with a fix, let me know if it still complains about anything that I didn't notice.

**Comment by jhogberg** [comment]
The tests ran fine over the weekend, so I'll merge the PR if you're happy with it.

**Comment by igors** [comment]
> The tests ran fine over the weekend, so I'll merge the PR if you're happy with it.

Sounds good to me, thanks!

**Comment by jhogberg** [comment]
Great, thanks again for the PR!

**Pull request merged**